### PR TITLE
[23149] Use wpCacheService to update form/field/display-attribute when resource changed

### PR DESCRIPTION
**Pull request by oliverguenther**
Instead of using watchers and several observers on the wpCacheService, we can use the form controller as the instrumenter. It replays the observed changes from the cache services to registered observers (namely `wp-edit-fields` and `display-attrs`), which can react according to the changes (not necessarily always changing the scope doing so).

https://community.openproject.com/work_packages/23149/activity

### CODE REVIEW

**Comment by TeatroIO** [comment]
I've prepared a stage to preview changes. <a href='http://oliverguenther-fix-23149-work-package-updates.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/' target='_blank'>Open stage</a> or <a href='http://oliverguenther-fix-23149-work-package-updates.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com//teatro.log' target='_blank'>view logs</a>.

**Comment by ulferts** [comment]
@oliverguenther I don't think this quite works. When switching the type of a work package to a milestone type, the date fields do not get updated. I believe that the schema is not fetched correctly.

File: frontend/app/components/wp-edit/wp-edit-field/wp-edit-field.directive.ts [hunk 1]
```diff
@@ -218,10 +225,17 @@ function wpEditFieldLink(scope,
...
                          controllers: [WorkPackageEditFormController, WorkPackageEditFieldController]) {
 
-  controllers[1].formCtrl = controllers[0];
-  controllers[1].formCtrl.registerField(scope.vm);
+  var formCtrl = controllers[0];
+  controllers[1].formCtrl = formCtrl;
+
+  formCtrl.registerField(scope.vm);
+  formCtrl.onWorkPackageUpdated('wp-edit-field-' + scope.vm.fieldName, (wp) => {
+    scope.vm.initializeField(wp);
```
Comment (ulferts) [inline frontend/app/components/wp-edit/wp-edit-field/wp-edit-field.directive.ts]: I think we should switch to a getter/setter for `this.work_package` and set the value here before calling `initializeField` (then not needing to pass the `wp` any more). With the change at hand we now have two ways in which a work package is passed to the directive:
1) via the watcher defined here
2) via the `workPackage` getter accessing the form's workPackage

We can assume the two methods to deliver the same results but that is only an assumption.

**Comment by oliverguenther** [comment]
This PR predates the merge of your date fix, but I believe the schema fetching actually works correctly. Try adding a new required custom field in the Milestone type and it is requested appropriately.

**Comment by oliverguenther** [comment]
@ulferts I've resolved the setter comment and gave this another whirl after the rebase. Switching to milestone shortly displays `invalid date`  (there's another ticket for that) and correctly updates it after that. If you can confirm this, I'd suggest to merge this and address the `evalAsync` remark from Roman separately.

**Comment by ulferts** [comment]
@oliverguenther for https://community.openproject.com/work_packages/23071/activity I will rework the wp-display-attr directive which also fixes the `invalid date` problems.

**Comment by ulferts** [comment]
Looks good so far @oliverguenther but there is one minor bug did not happen beforehand but that might just have been because the schema was not altered so that the fields did not change. The bug happens when changing a work package from a milestone type to a non milestone type. If the user has the "due date"/"start date" columns displayed in the table, those columns will become empty even if they have a value. 

Let me know if you think this should be addressed in a separate PR. I am willing to merge this PR as it is because without it we have other bugs on the mentioned fields.

**Comment by oliverguenther** [comment]
I've noticed the same issue when switching back and forth from a Type with a custom field. Previously, we had this unclear state of persistence of a value that was only valid for the old type. On the PR, I've first noticed that switching back no longer exposes that value. I don't think I've explicitly touched that area in my changes, but couldn't quite figure out where that stems from.

**Pull request merged**